src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -45,26 +50,112 @@ public HeaderModelBinder(ILoggerFactory loggerFactory)
+            _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
+        }
+
+        // to enable unit testing
+        internal IModelBinder InnerModelBinder { get; }
+
         /// <inheritdoc />
-        public Task BindModelAsync(ModelBindingContext bindingContext)
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext == null)
             {
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            var request = bindingContext.HttpContext.Request;
+            _logger.AttemptingToBindModel(bindingContext);
 
             // Property name can be null if the model metadata represents a type (rather than a property or parameter).
             var headerName = bindingContext.FieldName;
 
-            _logger.AttemptingToBindModel(bindingContext);
-
+            var request = bindingContext.HttpContext.Request;
             if (!request.Headers.ContainsKey(headerName))
             {
                 _logger.FoundNoValueInRequest(bindingContext);
             }
 
+            if (InnerModelBinder == null)
+            {
+                BindWithoutInnerBinder(bindingContext);
+                return;
+            }
+
+            // Do not set ModelBindingResult to Failed on not finding the value in the header as we want the inner 
+            // modelbinder to do that. This would give a chance to the inner binder to add more useful information.
+            // For example, SimpleTypeModelBinder adds a model error when binding to let's say an integer and the
+            // model is null.
+
+            var headerValueProvider = GetHeaderValueProvider(headerName, bindingContext);
+
+            // Create a new binding scope in order to supply the HeaderValueProvider so that the binders like
+            // SimpleTypeModelBinder can find values from header.
+            ModelBindingResult result;
+            using (bindingContext.EnterNestedScope(
+                    bindingContext.ModelMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: bindingContext.ModelName,
+                    model: bindingContext.Model))
+            {
+                bindingContext.ValueProvider = headerValueProvider;
+
+                await InnerModelBinder.BindModelAsync(bindingContext);
+                result = bindingContext.Result;
+            }
+
+            bindingContext.Result = result;
+
+            _logger.DoneAttemptingToBindModel(bindingContext);
+        }
+
+        private HeaderValueProvider GetHeaderValueProvider(string headerName, ModelBindingContext bindingContext)
+        {
+            var request = bindingContext.HttpContext.Request;
+
+            // Prevent breaking existing users in scenarios where they are binding to a 'string' property
+            // and expect the whole comma separated string, if any, as a single string and not as a string array.
```

Comment:
Thought we had code that basically said "if mapping a value provider collection to a `string`, concatenate the collection's items"? If the difference is only between `"X,Y,Z"` and `"X, Y, Z"`, we shouldn't bother with this extra bit.

Reply:
> Thought we had code that basically said "if mapping a value provider collection to a string, concatenate the collection's items"? If the difference is only between "X,Y,Z" and "X, Y, Z", we shouldn't bother with this extra bit.

The difference is between "X,Y,Z" and "X" (i.e if we always get comma separated values, the STMB would only take the first value in the list of values to bind to a string)

> Have we considered using a compat switch to drive this new model binding behavior? We wouldn't need this check at this point

Good point. @dougbu what do you think? If you agree, please give a name for the switch too :-)

Reply:
@dougbu I am interested in this switch because it would give a consistent experience (for example, if a query string had multiple values for a key, we only pick the first one).

Reply:
1. The "entire header string" behaviour is what we want. Leave it alone because headers aren't query string values.
2. A switch to control whether or not the HMB uses the old code path despite a non-`null` inner binder might be interesting if anything behaved differently for the previously supported cases i.e. when binding `string` or collections of `string`. Is anything **visible** changing there? I'm hoping not; any visible behaviour change would be unintended.
3. The other possibility would be to disable binding non-`string`s / non-`string` collections when a `CompatibitySwitch` is set. That might be useful because someone may have implemented a model binding provider that handles, say, `int`s when the `BindingSource` is `Header`. @Eilon should we add a switch for this fairly-extreme customization?

Reply:
> Is anything visible changing there? I'm hoping not; any visible behaviour change would be unintended.

I have tests covering the old behavior(2.0) using a non-null InnerBinder. So we are fine.
https://github.com/aspnet/Mvc/pull/7294/files#diff-7e7d3101aa01d6dab9e9b49c4c3ddf51R132

Reply:
If we think it's the right thing for new 2.1 apps to get a behavior that's different from 2.0, then, yeah, add a compat switch.